src/java.base/share/classes/java/util/concurrent/LinkedBlockingDeque.java

```diff
@@ -334,11 +330,16 @@ public void addLast(E e) {
      */
     public boolean offerFirst(E e) {
         if (e == null) throw new NullPointerException();
+        if (count >= capacity)
+            return false;
         Node<E> node = new Node<E>(e);
         final ReentrantLock lock = this.lock;
         lock.lock();
         try {
-            return linkFirst(node);
+            if (count >= capacity)
+                return false;
+            linkFirst(node);
```

Comment:
I'm a bit uneasy about incrementing the `count` in `linkFirst` but not enforcing the invariant. What's the benefit to changing linkFirst and linkLast to return void instead of keeping the original returning a boolean?

Reply:
> I'm a bit uneasy about incrementing the `count` in `linkFirst` but not enforcing the invariant. What's the benefit to changing linkFirst and linkLast to return void instead of keeping the original returning a boolean?

I based the approach on the LBQ enqueue() and dequeue() methods, which also return void and have a comment with the assertion.

Reply:
I understand, alas LBQ enqueue() and dequeue() don't increment a counter. It seems a bit brittle that the invariant needs to be maintained *externally* to the state modification. For the sake of maintainability I think the checks and the increment/decrement belongs in the same unit of work.

Reply:
This is what I meant about reducing sequential consistency. But it does work out in terms of maintaining specs.

Reply:
@kabutz I'd think maintaining the invariants within linkFirst and linkLast would be preferable (`count` must be re-read under the lock anyway)

Reply:
What would you like to do if the invariant fails inside linkFirst() and linkLast()? Should we throw an AssertionError each time?

Reply:
No, I was more thinking keeping it as it was: (return true/false from linkFirst / linkLast depending on whether it passes the capacity check or not)

Reply:
I've made the changes you suggested and it is indeed better. Thanks @viktorklang-ora